tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -154,6 +157,139 @@ public static void fastStartSpan(String operation) {
         fastStartSpan(operation, SpanType.LOCAL);
     }
 
+    /**
+     * Like {@link #startSpan(String, SpanType)}, but does not set or modify tracing thread state.
+     * This is an internal utility that should not be called directly outside of {@link DetachedSpan}.
+     */
+    static DetachedSpan detachInternal(String operation, SpanType type) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        String traceId = maybeCurrentTrace != null
+                ? maybeCurrentTrace.getTraceId() : Tracers.randomId();
+        boolean sampled = maybeCurrentTrace != null
+                ? maybeCurrentTrace.isObservable() : sampler.sample();
+        return sampled
+                ? new SampledDetachedSpan(operation, type, traceId, getParentSpanId(maybeCurrentTrace))
+                : new UnsampledDetachedSpan(traceId);
+    }
+
+    private static Optional<String> getParentSpanId(@Nullable Trace trace) {
+        if (trace != null) {
+            Optional<OpenSpan> maybeOpenSpan = trace.top();
+            if (maybeOpenSpan.isPresent()) {
+                return Optional.of(maybeOpenSpan.get().getSpanId());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static final class SampledDetachedSpan implements DetachedSpan {
+
+        private final AtomicBoolean completed = new AtomicBoolean();
+        private final String traceId;
+        private final OpenSpan openSpan;
+
+        SampledDetachedSpan(
+                String operation, SpanType type, String traceId, Optional<String> parentSpanId) {
+            this.traceId = traceId;
+            this.openSpan = OpenSpan.builder()
+                    .parentSpanId(parentSpanId)
+                    .spanId(Tracers.randomId())
+                    .operation(operation)
+                    .type(type)
+                    .build();
+        }
+
+        @Override
+        public SpanToken attach(String operationName, SpanType type) {
```

Comment:
Do you think it would be more ergonomic to "complete" the detached span when attaching it? I think this would encourage using the detachedSpan to indicate the amount of time spent between workers. i.e. some producers putting detachedSpans on queue somewhere and then having consumers pop them off and attach them.

Reply:
It would absolutely be more ergonomic, however it would not solve a set of the cases that this model is meant to address.
A standard attached span can have multiple children, I think it's similarly helpful to support multiple children using detached spans.

Given the DetachedSpan API, we can build ergonomic helpers which do precisely what you have suggested.